IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,466 @@
+  - Schema:
+https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?path=%2FUnifiedSchema%2FSchema.md&version=GBdev
+
+*/
+
+@interface MSIDMacKeychainTokenCache ()
+@end
+
+@implementation MSIDMacKeychainTokenCache
+
+#pragma mark - init
+
+- (id)init {
+    self = [super init];
+    return self;
+}
+
+#pragma mark - Accounts
+
+// Write an account to the macOS keychain cache.
+- (BOOL)saveAccount:(MSIDAccountCacheItem *)account
+                key:(MSIDCacheKey *)key
+         serializer:(id<MSIDAccountItemSerializer>)serializer
+            context:(id<MSIDRequestContext>)context
+              error:(NSError **)error {
+    MSID_TRACE;
+    // Get previous account, so we don't lose any fields
+    NSError* readError;
+    MSIDAccountCacheItem *previousAccount = [self accountWithKey:key serializer:serializer context:context error:&readError];
+    if (readError) {
+        if (error) {
+            *error = readError;
+        }
+        return FALSE;
+    }
+    if (previousAccount) {
+        [account updateFieldsFromAccount:previousAccount];
+    }
+
+    NSData *jsonData = [serializer serializeAccountCacheItem:account];
+    if (!jsonData) {
+        NSString *errorMessage = @"Failed to serialize account to json data.";
+        MSID_LOG_WARN(context, @"%@", errorMessage);
+        if (error) {
+            *error = MSIDCreateError(
+                MSIDErrorDomain, (NSInteger)MSIDErrorInternal, errorMessage, nil, nil, nil, context.correlationId, nil);
+        }
+        return FALSE;
+    }
+
+    NSDictionary *query = [self defaultAccountQuery:key];
+    NSDictionary *update = @{(id)kSecValueData: jsonData};
+
+    OSStatus status;
+    if (previousAccount) {
+        status = SecItemUpdate((CFDictionaryRef)query, (CFDictionaryRef)update);
+    } else {
+        NSMutableDictionary *dict = [query mutableCopy];
+        [dict addEntriesFromDictionary:update];
+        status = SecItemAdd((CFDictionaryRef)dict, NULL);
+    }
+
+    if (status != errSecSuccess) {
+        NSString *errorMessage = @"Failed to write account to keychain";
+        MSID_LOG_WARN(context, @"%@ (%d)", errorMessage, status);
+        if (error) {
+            *error = MSIDCreateError(
+                NSOSStatusErrorDomain, (NSInteger)status, errorMessage, nil, nil, nil, context.correlationId, nil);
+        }
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+// Read a single account from the macOS keychain cache.
+// If multiple matches are found, return nil and set an error.
+- (MSIDAccountCacheItem *)accountWithKey:(MSIDCacheKey *)key
+                              serializer:(id<MSIDAccountItemSerializer>)serializer
+                                 context:(__unused id<MSIDRequestContext>)context
+                                   error:(NSError **)error {
+    MSID_TRACE;
+
+    NSArray<MSIDAccountCacheItem *> *items = [self accountsWithKey:key
```

Comment:
This reads all accounts from the keychain. Can this be more efficient?

Reply:
I added same comment :)

Reply:
I shared the same concern, improvements checked in